src/protocol/v911s_nrf24l01.c

```diff
@@ -0,0 +1,313 @@
+};
+
+// flags going to packet[1]
+#define V911S_FLAG_EXPERT   0x04
+// flags going to packet[2]
+#define V911S_FLAG_CALIB    0x01
+
+// For code readability
+enum {
+    CHANNEL1 = 0,
+    CHANNEL2,
+    CHANNEL3,
+    CHANNEL4,
+    CHANNEL5,  
+};
+#define CHANNEL_CALIB   CHANNEL5
+
+// Bit vector from bit position
+#define BV(bit) (1 << bit)
+
+#define CHAN_RANGE (CHAN_MAX_VALUE - CHAN_MIN_VALUE)
+static u16 scale_channel(u8 ch, u16 destMin, u16 destMax)
+{
+    s32 chanval = Channels[ch];
+    s32 range = (s32) destMax - (s32) destMin;
+
+    if (chanval < CHAN_MIN_VALUE)
+        chanval = CHAN_MIN_VALUE;
+    else if (chanval > CHAN_MAX_VALUE)
+        chanval = CHAN_MAX_VALUE;
+    return (range * (chanval - CHAN_MIN_VALUE)) / CHAN_RANGE + destMin;
+}
+
+#define GET_FLAG(ch, mask) (Channels[ch] > 0 ? mask : 0)
+static void V911S_send_packet(u8 bind)
+{
+    u8 channel=hopping_frequency_no;
+    if(bind)
+    {
+        packet[0] = 0x42;
+        packet[1] = 0x4E;
+        packet[2] = 0x44;
+        for(u8 i=0;i<5;i++)
+            packet[i+3] = rx_tx_addr[i];
+        for(u8 i=0;i<8;i++)
+            packet[i+8] = hopping_frequency[i];
+    }
+    else
+    {
+        if(rf_ch_num&1)
+        {
+            if((hopping_frequency_no&1)==0)
+                channel+=8;
+            channel>>=1;
+        }
+        if(rf_ch_num&2)
+            channel=7-channel;
+        packet[ 0]=(rf_ch_num<<3)|channel;
+        packet[ 1]=V911S_FLAG_EXPERT; // short press on left button
+        packet[ 2]=GET_FLAG(CHANNEL_CALIB,V911S_FLAG_CALIB); // long  press on right button
+        memset(packet+3,0x00,14);
+        //packet[3..6]=trims TAER signed
+        u16 ch=scale_channel(CHANNEL3 ,0,0x7FF);// throttle
+        packet[ 7] = ch;
+        packet[ 8] = ch>>8;
+        ch=scale_channel(CHANNEL1 ,0x7FF,0);    // aileron
+        packet[ 8]|= ch<<3;
+        packet[ 9] = ch>>5;
+        ch=scale_channel(CHANNEL2,0,0x7FF);     // elevator
+        packet[10] = ch;
+        packet[11] = ch>>8;
+        ch=scale_channel(CHANNEL4,0x7FF,0);      // rudder
+        packet[11]|= ch<<3;
+        packet[12] = ch>>5;
+    }
+    
+    // Power on, TX mode, 2byte CRC
+    XN297_Configure(BV(NRF24L01_00_EN_CRC) | BV(NRF24L01_00_CRCO) | BV(NRF24L01_00_PWR_UP));
+    if (!bind)
+    {
+        NRF24L01_WriteReg(NRF24L01_05_RF_CH, hopping_frequency[channel]);
+        hopping_frequency_no++;
+        hopping_frequency_no&=7;    // 8 RF channels
+    }
+    // clear packet status bits and TX FIFO
+    NRF24L01_WriteReg(NRF24L01_07_STATUS, 0x70);
+    NRF24L01_FlushTx();
+    XN297_WritePayload(packet, V911S_PACKET_SIZE);
+
+    if (tx_power != Model.tx_power) {
+        //Keep transmit power updated
+        tx_power = Model.tx_power;
+        NRF24L01_SetPower(tx_power);
+    }
+}
+
+static void V911S_init()
+{
+    NRF24L01_Initialize();
+    NRF24L01_SetTxRxMode(TX_EN);
+    XN297_SetTXAddr((u8 *)"\x4B\x4E\x42\x4E\x44", 5);          // Bind address
```

Comment:
Does this prevent us having two people control two planes in the same model?

Reply:
no, that's only the bind address, it's common to every transmitters.